### PR TITLE
8255237: ZGC: Bulk free garbage pages during relocation set selection

### PR DESCRIPTION
**Pull request by pliden**
During relocation set selection, we immediately free garbage (empty) pages. On very large heap, which has a very high number of garbage pages we can make a significant number of calls to free pages, which prolongs the "Concurrent Select Relocation Set" phase. Freeing a page involves grabbing a lock, inserting the page into the page cache, and then try to satisfy any stalled allocation. By freeing empty pages in bulk, we can amortize the some of the cost of freeing a page and speed up this phase. With this patch we free (up to) 64 pages in bulk.

This PR depends on "8255234: ZGC: Bulk allocate forwarding data structures" and "8255235: ZGC: Allocate and initialize forwarding data structures in parallel", which are separate PRs (#804 and #805).

This patch was joint work with Albert (ayang).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8255237](https://bugs.openjdk.java.net/browse/JDK-8255237): ZGC: Bulk free garbage pages during relocation set selection

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - Author) ⚠️ Review applies to 29d24bbe65969020e1f130a8d5cb330555e185e0
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to 91cfa738563d32ca3ce3168729d46a94189bdcfe

### Contributors
 * Albert Mingkun Yang `<ayang@openjdk.org>`
 * Per Liden `<pliden@openjdk.org>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/806/head:pull/806`
`$ git checkout pull/806`

### CODE REVIEW

File: src/hotspot/share/gc/z/zHeap.cpp [hunk 1]
```diff
@@ -369,11 +388,14 @@ void ZHeap::select_relocation_set() {
       // Register garbage page
       selector.register_garbage_page(page);
 
-      // Reclaim page immediately
-      free_page(page, true /* reclaimed */);
+      // Reclaim garbage pages in bulk
```
Comment (albertnetymk) [inline src/hotspot/share/gc/z/zHeap.cpp:391]: Maybe adding a comment that this is an optimization?

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/z/zHeap.cpp (hunk 1 above)
Comment (pliden) [reply src/hotspot/share/gc/z/zHeap.cpp:391]: Will fix.

Comment (albertnetymk) [reply src/hotspot/share/gc/z/zHeap.cpp:394]: Thank you.

**Review by fisk — APPROVED** [review]
A bit tricky to review when it contains more than one patch. But looks good as far as I can see.

**Comment by pliden** [comment]
@fisk Thanks for reviewing! Yes, eagerly awaiting the Skara support for better handling of dependent PRs. In this case, looking at the original commit and (https://github.com/openjdk/jdk/pull/806/commits/29d24bbe65969020e1f130a8d5cb330555e185e0) and the follow commit up to Albert's comment (https://github.com/openjdk/jdk/pull/806/commits/132efd58548602558e34482a75c31fa7e2b889f3) is probably the easiest way to review this.

**Pull request closed without merging**